tests/schema/tests.py

```diff
@@ -2391,6 +2395,338 @@ def test_text_field_with_db_index(self):
         assertion = self.assertIn if connection.features.supports_index_on_text_field else self.assertNotIn
         assertion('text_field', self.get_indexes(AuthorTextFieldWithIndex._meta.db_table))
 
+    @skipUnlessDBFeature('supports_expression_indexes')
+    def test_func_index(self):
+        """
+        Tests creating a functional index
+        """
+
+        # Create the table
+        with connection.schema_editor() as editor:
+            editor.create_model(Author)
+            editor.create_model(Book)
+
+        # Define the index
+        index_name = 'lower_func_idx'
+        index = Index(fields=[Lower('title').desc()], name=index_name)
+
+        # Add the index
+        with connection.schema_editor() as editor:
+            editor.add_index(Book, index)
+            sql = index.create_sql(Book, editor)
+        self.assertIn(index_name, self.get_constraints(Book._meta.db_table))
+
+        sql = str(sql).upper()
+        self.assertIn('LOWER(', sql)
```

Comment:
Could `Columns` be used here?

Reply:
Couldn't any examples of usages of the DDL in other tests. One usage I could find would be something like:
```python
self.assertTrue(sql.references_column('schema_book', '(LOWER("title"))'))
```

This requires to specify the fully quoted and wrapped function which makes it hard to write tests which works on all engines.

Another better usage could be:
```python
self.assertIn('DESC', sql.parts['columns'].col_suffixes)
```

Let me know if you have any ideas how the DDL could be better applied.

Reply:
You have `quote_name` and `quote_value` so you don't need to write vendor specific quotes. Anyway, it was meant as a thought, because I think it looks a bit weird with those sequences of `x < y`.

Reply:
Maybe `get_constraints['columns']` could be used for some tests?

Reply:
I agree that it does look a bit weird. I've tried using `get_constraints['columns']` but unfortunately none of the implementations support functional indexes at the moment. A `None` value is returned for functional indexes. 

```python
Index(fields=["email", Lower("name")], name="foo_idx")
```

A call to get_constraints will look like this:
```python
>>> self.get_constraints(Foo._meta.db_table))["foo_idx"]["columns"]
['email', None]
```

MySQL, PostgresSQL and Sqlite all have the same behaviour. I guess it would be beneficial to add support for functional indexes in `get_constraints`. For MySQL it would be pretty straightforward. In `SHOW INDEX` it's just a new column, see:
https://dev.mysql.com/doc/refman/8.0/en/show-index.html 

> Expression
> MySQL 8.0.13 and higher supports functional key parts (see Functional Key Parts), which affects both the Column_name and Expression columns:
> For a nonfunctional key part, Column_name indicates the column indexed by the key part and Expression is NULL.
> For a functional key part, Column_name column is NULL and Expression indicates the expression for the key part.

For Postgres it's a bit tricker. I haven't really figured out how to adjust the query.

For SQLite I'm not even sure if it's possible to retrieve the column name:
https://www.sqlite.org/pragma.html#pragma_index_xinfo

> 3. The name of the column being indexed, or NULL if the index-column is the rowid of the table being indexed or an expression.

Reply:
OK. I'm fairly sure that this will need to be implemented for this patch to be merged. I had a look at PostgreSQL last night, and so far just read about `pg_get_expr` from https://www.postgresql.org/docs/current/functions-info.html I'll experiment to see if I can help you with the query today.

Reply:
https://www.postgresql.org/message-id/6721.1357856188%40sss.pgh.pa.us So probably the whole debate about touching the introspection for expressions should be closed. I could only get `pg_get_expr` to output whatever I used for `CREATE INDEX .. ON (lower(body), lower(title))` where `lower(body), lower(title)` would be returned by `pg_get_expr`.

Reply:
Yes I agree, I think it'll be hard to get consistent behaviour across all database engines. It looks like SQLite simply does not return anything comparable when using expressions. 

I had a look at how other ORM:s solve the index introspection for SQLite:

 ActiveRecord uses a regexp to parse the `CREATE INDEX` statement loosing all order of the actual index:
https://github.com/rails/rails/blob/master/activerecord/lib/active_record/connection_adapters/sqlite3/schema_statements.rb#L24

SQLAlchemy simply ignores these kinds of indexes for introspection: https://github.com/sqlalchemy/sqlalchemy/blob/master/lib/sqlalchemy/dialects/sqlite/base.py#L2039